include/oneapi/dpl/internal/distributed_ranges_impl/sp/init.hpp

```diff
@@ -0,0 +1,166 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cassert>
+#include <memory>
+#include <span>
+#include <sycl/sycl.hpp>
+#include <type_traits>
+#include <vector>
+
+#include <oneapi/dpl/execution>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/execution_policy.hpp>
+#include <oneapi/dpl/internal/distributed_ranges_impl/sp/util.hpp>
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+namespace __detail
+{
+
+inline sycl::context* global_context_;
+
+inline std::vector<sycl::device> devices_;
+
+inline std::vector<sycl::queue> queues_;
+
+inline std::vector<oneapi::dpl::execution::device_policy<>> dpl_policies_;
+
+inline std::size_t ngpus_;
+
+inline sycl::context&
+global_context()
+{
+    return *global_context_;
+}
+
+inline std::size_t
+ngpus()
+{
+    return ngpus_;
+}
+
+inline std::span<sycl::device>
+global_devices()
+{
+    return devices_;
+}
+
+} // namespace __detail
+
+inline sycl::context&
+context()
+{
+    return __detail::global_context();
+}
+
+inline std::span<sycl::device>
+devices()
+{
+    return __detail::global_devices();
+}
+
+inline std::size_t
+nprocs()
+{
+    return __detail::ngpus();
+}
+
+inline distributed_device_policy par_unseq;
```

Comment:
This naming is very confusing, especially when the policy is used only by its name, without any namespaces. I strongly recommend to rename - e.g., as `sycl_device_collection` which is what any `distributed_device_policy` really is.

Reply:
I was just a bit late with my comment to the PR that tries to address the above,, so I will add it here as well.

My comment above is admittedly not quite clear, sorry if it mislead you. But I actually meant the par_unseq name of the object, not the type. That is, I do not like `par_unseq` as the policy name, finding it misleading.

I can see why you may want it to make the use of DR algorithms similar to the standard ones. But this is only syntactic similarity, while semantically the differences are rather broad. Also, I am afraid it will get confused with `oneapi::dpl::execution::par_unseq` which closely follows the standard semantics.

What do you think about `struct device_set_policy` and `inline device_set_policy preset_devices`?

@mateuszpn @lslusarczyk @BenBrock @rscohn2